# Non-Linear Programming Example using Python and Scipy

Non-linear programming is a mathematical optimization technique used to optimize non-linear objective functions subject to non-linear constraints. Non-linear programming is useful in many real-world applications, such as finance, engineering, and computer science.

In this Jupyter Notebook, I showcase an example of situation where non-linear programming can be applied and how we can use the Python library Scipy to optimize the results.

## Portfolio Optimization using the Black-Litterman Model

In portfolio optimization, the goal is to find the optimal weights for a portfolio of assets that maximizes revenue while taking into account risk and other constraints. The Black-Litterman model is a popular approach to portfolio optimization that combines market equilibrium with an investor's views on the market to generate optimal portfolio weights.

### Variables

Here are the variables used in the code:

| Variable | Description |
| -------- | ----------- |
| $P$ | The views matrix |
| $q$ | The return vector |
| ${\Omega}$ | The uncertainty matrix |
| ${\tau}$ | The risk aversion parameter |
| ${\Sigma}$ | The covariance matrix |
| ${\pi}$ | The equilibrium excess returns vector |
| ${\mu}$ | The optimal portfolio weights |
| ${\bar\mu}$ | The expected returns vector for the assets |
| `obj` | The objective function for the optimization problem |
| `con_target` | The constraint function for the target expected return |
| `con_budget` | The constraint function for the budget constraint |
| `x0` | The initial guess for the portfolio weights |
| `result` | The optimal portfolio weights calculated using the SLSQP method |

**Note:** In Black-Litterman model, the expected returns of assets are modeled as a combination of the prior expected returns (mu) and the views of the investor (Q). However, the prior expected returns (mu) are often based on historical data and may not be accurate or reflective of current market conditions. Therefore, Black-Litterman model introduces an additional constraint in the optimization problem to ensure that the portfolio's expected return is greater than or equal to a specified minimum expected return (${\bar\mu}$).

By including ${\bar\mu}$ as a constraint, Black-Litterman model helps to ensure that the resulting portfolio is not only diversified and efficient, but also meets the investor's minimum return requirement. This is particularly important for risk-averse investors who want to ensure that their portfolio generates sufficient returns to meet their financial goals.

### Objective Function

The objective function is based on the covariance between the stocks, bonds, and money market fund:

**Minimize:** 

$$ 0.02778*x_{S}^2 + 2 * (0.00387)*x_{S}x_{B} \\
+ 2 * (0.00021)*x_{S}x_{M} + 0.01112*x_{B}^2 \\
- 2 * (0.00020)*x_{B}x_{M} + 0.00115*x_{M}^2 $$

where:
- $x_{S}$: Portfolio weight for stock, denoted as x[0] in Python code
- $x_{B}$: Portfolio weight for bonds, denoted as x[1] in Python code
- $x_{M}$: Portfolio weight for money market, denoted as x[2] in Python code

The objective function is based on the covariance between the three assets. Here is the covariance table:

|        | Stocks | Bonds  | MM     |
|--------|--------|--------|--------|
| Stocks | 0.02778| 0.00387| 0.00021|
| Bonds  | 0.00387| 0.01112|-0.00020|
| MM     | 0.00021|-0.00020| 0.00115|


### Constraints

Here is a table of the mean rate of return ${\bar\mu}$ for each asset:

|                            | Stocks  | Bonds   | MM      |
|----------------------------|---------|---------|---------|
| Mean rate of return of ${\bar\mu}$ | 11.77%  | 7.51%   | 2.34%   |


In mathematical notation, the constraints can be represented as:
$\sum\limits_{i=1}^{3} x_i = 1$ | {
where $\mathbf{x}$ is the vector of portfolio weights}

$\boldsymbol{\mu}^T \mathbf{x} = R$ | {
$\boldsymbol{\mu}$ is the The optimal portfolio weights of mean rates of return, and $R$ is the minimum expected return}

### Calculating ${\mu}$

To calculate the optimal portfolio weights using the Black-Litterman model, the following steps can be taken:

Define the views matrix P, the return vector q, and the uncertainty matrix Omega. The matrix P represents the views on the asset returns and is a 2x3 matrix. The return vector q represents the expected return of the views and is a 2x1 matrix. The uncertainty matrix Omega represents the uncertainty of the views and is a 2x2 diagonal matrix.

$P$ =
$\begin{bmatrix} 0 &  0 & 1 \\ 1 & -1 & 0\end{bmatrix}$

$q$ =
$\begin{bmatrix} 0.02 \\ 0.05 \\ 0 \end{bmatrix}$

${\Omega}$ =
$\begin{bmatrix} 0.00001 & 0 \\ 0 & 0.001\end{bmatrix}$


Define the risk aversion parameter tao and the covariance matrix Sigma. The covariance matrix Sigma is a 3x3 matrix that represents the covariance between the asset returns.

${\tau}$ = = 0.1

${\Sigma}$ =
$\begin{bmatrix} 0.02778 & 0.00387 & 0.00021 \\ 0.00387 & 0.01112 & -0.0002 \\ 0.00021 & -0.0002 & 0.00115 \end{bmatrix}$

Compute the equilibrium excess returns vector pi. The equilibrium excess returns vector pi is a 3x1 vector that represents the expected returns of the assets given their risk.

${\pi}$ =
$\begin{bmatrix} 01073 &  0.0737 & 0.0627\end{bmatrix}$

Compute the optimal portfolio weights ${\mu}$ using the Black-Litterman model. The optimal portfolio weights ${\mu}$ is a 3x1 vector that represents the weights of the assets in the optimal portfolio.

The formula for ${\mu}$ is:

$\mu = \left( \text{inv}\left(\tau \Sigma^{-1} + P^\top \Omega^{-1} P \right) \right) \left( \tau \Sigma^{-1} \pi + P^\top \Omega^{-1} q \right)$


In [16]:
from math import log, exp, inf
from scipy.optimize import minimize
import numpy as np
from numpy.linalg import inv

# Define the views matrix P, the return vector q, and the uncertainty matrix Omega
P = np.array([[0,0,1],[1, -1, 0]])
q = np.array([0.02, 0.05])
Omega = np.array([[0.00001,0],[0,0.001]])

# Define the risk aversion parameter tao and the covariance matrix Sigma
tao = 0.1
Sigma = np.array([[0.02778,0.00387,0.00021],[0.00387,0.01112,-0.0002],[0.00021,-0.0002,0.00115]])

# Compute the equilibrium excess returns vector pi
pi = np.array([0.1073,0.0737,0.0627])

# Compute the optimal portfolio weights mu using the Black-Litterman model
mu = inv(inv(tao*Sigma)+P.T @ inv(Omega) @ P) @ (inv(tao*Sigma) @ pi + P.T @ inv(Omega) @ q)

# Print the optimal portfolio weights
print(mu)

# Define the expected returns vector for the assets, with digits dropped or rounded
muu = [0.1177, 0.0751, 0.0234]

# Define the objective function for the optimization problem, which maximizes revenue
def obj(x):
    return 0.02778*x[0]*x[0] + \
            2*0.00387*x[0]*x[1] + \
            2*0.00021*x[0]*x[2] + \
            0.01112*x[1]*x[1] - \
            2*0.00020*x[1]*x[2] + \
            0.00115*x[2]*x[2] 

# Define the constraint function for the target expected return
def con_target(x):
    return mu[0]*x[0] + mu[1]*x[1] + mu[2]*x[2] - 0.115

# Define the constraint function for the budget constraint
def con_budget(x):
    return x[0]+x[1]+x[2]-1

# Define the initial guess for the portfolio weights
x0 = [1, 0, 0]

# Solve the optimization problem using the SLSQP method, with bounds and constraints
result = minimize(obj, x0, 
                  method = 'SLSQP',
                  bounds=[(0.000000001,inf)]*3, 
                  constraints=(
                      {'type':'ineq', 'fun':con_target},
                      {'type':'eq', 'fun':con_budget})
                  )

# Print the optimal portfolio weights
print(result)

[0.11779598 0.07520736 0.02344031]
     fun: 0.02477487942948806
     jac: array([0.05242058, 0.00869194, 0.00036617])
 message: 'Optimization terminated successfully'
    nfev: 16
     nit: 4
    njev: 4
  status: 0
 success: True
       x: array([9.34349205e-01, 6.56507936e-02, 1.00000006e-09])


### Conclusion & Results

Based on the optimization results, we can conclude that the optimal portfolio weights for the given covariance matrix and mean rate of return are approximately 93.43% for stocks ($x_S$), 6.57% for bonds ($x_B$), and 0.0000001% for money market ($x_M$), subject to the constraint that the expected return of the portfolio is at least 11.5%.

It is important to note that the small weight given to the money market in the optimal portfolio may suggest a low-risk investment strategy. Additionally, the relatively high weight given to stocks may suggest that the investor is willing to take on more risk in order to achieve higher returns. However, the specific investment strategy should take into account the investor's individual risk tolerance, financial goals, and time horizon.